[feature]: Add Update Available Message to keploy

Akash-Singh04 opened the pull request:
## Related Issue
  - Added the update available message to Keploy CLI

Closes: #1485 

#### Describe the changes you've made
- Added Update Available message to notify users about new updates
- Message only shows up if current version of keploy is not updated 
- Message does not show up on dev versions

## Type of change

<!--
Example how to mark a checkbox:-
- [x] My code follows the code style of this project.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added
NIL

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
NIL

## Checklist:
<!--
Example how to mark a checkbox:-
- [x] My code follows the code style of this project.
-->
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if any)
![image](https://github.com/keploy/keploy/assets/114267538/c2caaead-da85-4e97-9ef8-483218f43d4b)

Akash-Singh04 commented:
What needs to be discussed is the delay that will happen(~2-3s) due to an Http request being made everytime keploy is run


There are two possible solutions:

- The update message is shown at the end of execution of the command

- The update message is shown only once a day.


Let us discuss on how this can be implemented.

Akash-Singh04 commented:
@charankamarapu The review and discussion for this pull request is pending. Kindly look into it at your earliest convenience

PranshuSrivastava commented on cmd/root.go:
```diff
@@ -279,6 +282,28 @@ func (r *Root) execute() {
+		releaseInfo, err := utils.GetLatestGitHubRelease()
+		if err != nil {
+			r.logger.Debug("Failed to fetch the latest release version", zap.Error(err))
+			return
```
why are we returning if we failed to fetch the latest release, keploy should still work isn't it? This will confuse the user as to why keploy is not running

PranshuSrivastava commented on main.go:
```diff
@@ -37,7 +37,6 @@ func main() {
 	utils.Version = version
 	if binaryToDocker := os.Getenv("BINARY_TO_DOCKER"); binaryToDocker != "true" {
 		fmt.Println(logo, " ")
-		fmt.Printf("version: %v\n\n", version)
```
why did you remove this?

Akash-Singh04 replied:
Refrence: https://github.com/keploy/keploy/issues/1488

PranshuSrivastava replied:
I don't see any comments on that issue, did you discuss this with any organisation member before implementing? We usually print the version along with logo because whenever someone faces a problem with Keploy, we ask for the logs, and there we can check what specific version is causing the issue.

Akash-Singh04 replied:
Yes this was prior discussed with @charankamarapu , It was supposed to be implemented as a part of #1376, but was asked to be removed as some discussions needed to be done before the implementation. 

Can we have a discussion over this on slack and decide whats the best course of action regarding the version message?

Akash-Singh04 replied:
> What needs to be discussed is the delay that will happen(~2-3s) due to an Http request being made everytime keploy is run
> 
> There are two possible solutions:
> 
> * The update message is shown at the end of execution of the command
> * The update message is shown only once a day.
> 
> Let us discuss on how this can be implemented.


This also needs to be discussed with the community

PranshuSrivastava replied:
I had a chat with him and I think we are now supposed to run keploy --version everytime we start Keploy. Is that correct? Please let me know when youa add that change

Akash-Singh04 replied:
Updated, Thank you for resolving

Akash-Singh04 replied:
As it turns out, Keploy automatically adds in a --version flag upon initialisation, which was customised in #1376. Instead of running keploy with --version command everytime, i think its best we just update the message to 
```Keploy version: 2-xyz```


Let me know if this is an optimal approach

PranshuSrivastava replied:
I think it was finaliased that the version flag should be run everytime because showing the version like this is not a good pracice so I think we should stick to that.

PranshuSrivastava commented on cmd/root.go:
```diff
@@ -279,6 +282,28 @@ func (r *Root) execute() {
+               ╰─────────────────────────────────────╯
+			   `
+			versionmsg := fmt.Sprintf(msg, currentVersion, releaseInfo.TagName, updatetext)
+			fmt.Printf(versionmsg)
```
shouldn't we also give an option to the user to update keploy? or atleast tell him how he can do so by mentioning the command here.

Akash-Singh04 replied:
In my opinion asking the user to update immediately would be a bad user experience. We can however mention the command here to do so.

Akash-Singh04 commented:
@PranshuSrivastava It has been added again

charankamarapu commented:
Hey @Akash-Singh04 please update the branch with main.

Akash-Singh04 commented:
Hey @charankamarapu , Branch has been updated.
I was wondering if there is a way we can run the update message only when a new terminal session is created? Do you know a way we can get that information in our CLI? I tried searching for some solutions but wasn't able to get anything that fits our need.

PranshuSrivastava commented:
Hey @Akash-Singh04 , the checks are failing, can you please take a look? Also, the update command that you added has not been working correctly for users. Even after they run keploy update, and then type Keploy again, it shows the same version.

Akash-Singh04 commented:
@PranshuSrivastava I am not sure about your bug report for the update command. I used used it to update from ```Keploy 2.0.0-alpha78``` to  ```Keploy 2.1.0-alpha2```. Here are the screenshots:
![image](https://github.com/keploy/keploy/assets/114267538/27032170-b9e3-4c59-8b1b-b53d4d6c7e25)
![image](https://github.com/keploy/keploy/assets/114267538/6d4f303e-b77a-4b53-a544-63544f01ee42)

re-Tick commented:
There is an issue with the closing `|`. Please have a look.
![Uploading image.png…]()

Akash-Singh04 commented:
@re-Tick @PranshuSrivastava Both of the requested changes have been made
- The update command is now refactored to work reliably on all platforms as done in v1 in #1325
- Fixed #1608 Keploy update command updates even on latest version 
- Added a function to calculate whitespaces so that the closing braces of the update message are always inline

PranshuSrivastava commented:
<img width="1389" alt="image" src="https://github.com/keploy/keploy/assets/37413698/dba81ee2-ae6f-470e-977e-9f2ccb04eba0">
There is still an error in the graphics, it is aligned towards the right too much. I tried running the update command again, and even though it said it had updated the keploy version, when i opened a new terminal tab, it showed the same version. Can you try to just simply download the one touch install and execute if, instead of updating the binary inside the "/usr/ocal/bin" directory?

Akash-Singh04 commented:
Will look into this

nehagup commented:
Hi @Akash-Singh04 this is great initiative. One good addition to `keploy update` that users suggested is - we prompt "keploy update available.. (y/n)" for the first time, however, if the user enters "no" we should just log the warning (update available) and not prompt every time.

Akash-Singh04 commented:
@PranshuSrivastava Your requested changes have been made

nehagup commented:
@Sonichigo please review
